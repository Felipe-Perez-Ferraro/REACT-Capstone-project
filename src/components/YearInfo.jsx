import React, { useEffect } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faAngleLeft } from '@fortawesome/free-solid-svg-icons';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import { fetchCompaniesThunk } from '../redux/companies/companiesSlice';
import { screenStyle } from '../tailwindStyles';

function YearInfo() {
  const goBack = <FontAwesomeIcon icon={faAngleLeft} />;
  const dispatch = useDispatch();
  const { id } = useParams();
  const { symbol } = useParams();
  const { companies } = useSelector((state) => state.companies);
  const { isFetched } = useSelector((state) => state.companies);
  const yearSelected = companies.find((company) => company.id === id);

  useEffect(() => {
    if (!isFetched) {
      dispatch(fetchCompaniesThunk(symbol));
    }
  }, [fetchCompaniesThunk, isFetched]);

  return (
    <section className="section bg-pink-900 py-4">
      <article className={`${screenStyle} flex flex-col gap-4`}>
        <Link className="text-slate-100 font-semibold pb-3 text-left" to={`/company/${symbol}`}>
          {goBack}
          {' '}
          Back to years
        </Link>
        <table className="table border w-full bg-pink-600">
          <thead className="border bg-pink-800">
            <tr>
              <th className="text-slate-100 font-semibold text-xl py-3">{id}</th>
            </tr>
          </thead>
          <tbody>
            <tr className="bg-pink-600 even:bg-pink-700">
              <td className="flex justify-between px-1 lg:px-2 xl:px-2 py-3">
                <span className="text-slate-100 font-semibold lg:text-lg xl:text-lg">Costs and expenses:</span>
                {' '}
                <span className="text-slate-100 font-semibold">{yearSelected.costAndExpenses}</span>
              </td>
            </tr>
            <tr className="bg-pink-600 even:bg-pink-700">
              <td className="flex justify-between px-1 lg:px-2 xl:px-2 py-3">
                <span className="text-slate-100 font-semibold lg:text-lg xl:text-lg">Cost of revenue:</span>
                {' '}
                <span className="text-slate-100 font-semibold">{yearSelected.costOfRevenue}</span>
              </td>
            </tr>
            <tr className="bg-pink-600 even:bg-pink-700">
              <td className="flex justify-between px-1 lg:px-2 xl:px-2 py-3">
                <span className="text-slate-100 font-semibold lg:text-lg xl:text-lg">Gross Profit:</span>
                {' '}
                <span className="text-slate-100 font-semibold">{yearSelected.grossProfit}</span>
              </td>
            </tr>
            <tr className="bg-pink-600 even:bg-pink-700">
              <td className="flex justify-between px-1 lg:px-2 xl:px-2 py-3">
                <span className="text-slate-100 font-semibold lg:text-lg xl:text-lg">Income Tax Expense:</span>
                {' '}
                <span className="text-slate-100 font-semibold">{yearSelected.incomeTaxExpense}</span>
              </td>
            </tr>
            <tr className="bg-pink-600 even:bg-pink-700">
              <td className="flex justify-between px-1 lg:px-2 xl:px-2 py-3">
                <span className="text-slate-100 font-semibold lg:text-lg xl:text-lg">Interest Income:</span>
                {' '}
                <span className="text-slate-100 font-semibold">{yearSelected.interestIncome}</span>
              </td>
            </tr>
            <tr className="bg-pink-600 even:bg-pink-700">
              <td className="flex justify-between px-1 lg:px-2 xl:px-2 py-3">
                <span className="text-slate-100 font-semibold lg:text-lg xl:text-lg">Net Income:</span>
                {' '}
                <span className="text-slate-100 font-semibold">{yearSelected.netIncome}</span>
              </td>
            </tr>
            <tr className="bg-pink-600 even:bg-pink-700">
              <td className="flex justify-between px-1 lg:px-2 xl:px-2 py-3">
                <span className="text-slate-100 font-semibold lg:text-lg xl:text-lg">Operating Expenses:</span>
                {' '}
                <span className="text-slate-100 font-semibold">{yearSelected.operatingExpenses}</span>
              </td>
            </tr>
            <tr className="bg-pink-600 even:bg-pink-700">
              <td className="flex justify-between px-1 lg:px-2 xl:px-2 py-3">
                <span className="text-slate-100 font-semibold lg:text-lg xl:text-lg">Revenue:</span>
                {' '}
                <span className="text-slate-100 font-semibold">{yearSelected.revenue}</span>
              </td>
            </tr>
          </tbody>
        </table>
      </article>
    </section>
  );
}

export default YearInfo;
