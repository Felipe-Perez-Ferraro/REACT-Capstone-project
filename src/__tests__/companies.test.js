import { configureStore } from '@reduxjs/toolkit';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import YearInfo from '../components/YearInfo';
import companiesReducer, {
  fetchCompaniesThunk,
} from '../redux/companies/companiesSlice';
import fetchApi from '../services/api';

jest.mock('../services/api');

describe('companiesSlice tests', () => {
  let store;
  beforeEach(() => {
    store = configureStore({
      reducer: {
        companies: companiesReducer,
      },
    });
  });

  it('Should fetch the financial API', async () => {
    const mockData = [
      {
        id: 2023,
        calendarYear: 2023,
        costAndExpenses: 4352345345,
        costOfRevenue: 3241234,
        grossProfit: 312412341234,
        incomeTaxExpense: 764574567,
        interestIncome: 45625462456,
        netIncome: 7864678674,
        operatingExpenses: 234523342,
        revenue: 43584839734,
      },
    ];
    fetchApi.mockResolvedValue(mockData);
    await store.dispatch(fetchCompaniesThunk('symbol'));

    const state = store.getState().companies;
    expect(state.isLoading).toBe(false);
    expect(state.isFetched).toBe(true);
    expect(state.error).toBe(false);
    expect(state.companies).toEqual(mockData);
  });

  it('Should render a financial information', async () => {
    const mockData = [
      {
        id: 2023,
        calendarYear: 2023,
        costAndExpenses: 4352345345,
        costOfRevenue: 3241234,
        grossProfit: 312412341234,
        incomeTaxExpense: 764574567,
        interestIncome: 45625462456,
        netIncome: 7864678674,
        operatingExpenses: 234523342,
        revenue: 43584839734,
      },
    ];
    fetchApi.mockResolvedValue(mockData);
    await store.dispatch(fetchCompaniesThunk('symbol'));

    const { yearInfo } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/AAPL/2022']}>
          <Routes>
            <Route path="/:symbol/:id" element={<YearInfo />} />
          </Routes>
        </MemoryRouter>
      </Provider>,
    );

    expect(yearInfo).toMatchSnapshot();
    expect(screen.getByTestId('costAndExpenses').textContent).toBe(
      'Costs and expenses:',
    );
    expect(screen.getByTestId('costOfRevenue').textContent).toBe(
      'Cost of revenue:',
    );
    expect(screen.getByTestId('grossProfit').textContent).toBe('Gross Profit:');
    expect(screen.getByTestId('incomeTaxExpense').textContent).toBe(
      'Income Tax Expense:',
    );
    expect(screen.getByTestId('interestIncome').textContent).toBe(
      'Interest Income:',
    );
    expect(screen.getByTestId('netIncome').textContent).toBe('Net Income:');
    expect(screen.getByTestId('operatingExpenses').textContent).toBe(
      'Operating Expenses:',
    );
    expect(screen.getByTestId('revenue').textContent).toBe('Revenue:');
  });
});
