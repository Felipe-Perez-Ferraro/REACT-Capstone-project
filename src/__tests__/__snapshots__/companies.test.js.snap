// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`companiesSlice tests Should render a financial information 1`] = `undefined`;
